dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/Compilation.java

```diff
@@ -81,7 +81,11 @@ public void run(Program program) {
             logger.warn("Skipped compilation: Program is already compiled to {}", program.getArch());
             return;
         }
-
+        // TODO: Refactor processors such that compiler is resolved
+        //  based on the source language and target
+        if (program.getFormat() == Program.SourceLanguage.SPV) {
+            compiler = new VisitorSpirvVulkan();
+        }
```

Comment:
Merging them with vulkan will likely be a problem when we add opencl support for spirv code. In this case, we will need tags of the opencl model and possibly other opencl-specific manipulations. For this, we will need a new spirv-to-opencl (internal) compilation.

I strongly believe that we should separate the original spirv events and their versions for vulkan and opencl models.

Reply:
But then you should create a new spirv target to make the difference between spirv and vulkan (and eventually opencl) events explicit. With this new target you can simple decide the compiler in `Compilarion.java` in the same way it is done for all other targets.

Reply:
"target" is everything that needs to be compiled/converted to base events ... C11/IMM/LKMM fit in the same category and we still have a target/Arch for them

Reply:
I don't exactly know what the issue is, but just for clarification of what Hernán said:
`target` is not necessarily a hardware target but can also be the language itself, if the language itself has a memory model we want to support. In this case, the internal compilation done by Dartagnan does not correspond to/mimic a real source-to-hardware compilation.
I don't know what the exact relation between SPIRV, Vulkan, and OpenCL is to tell what exactly needs to be though.
For example, is `spirv.cat` the Vulkan memory model or actually the SPIRV memory model (if these are even fundamentally different things?).

Reply:
spirv.cat is a wrong name, should be vulkan.cat

Reply:
The name of the file was intentional to match the [alloy model file](https://github.com/KhronosGroup/Vulkan-MemoryModel/blob/main/alloy/spirv.als), but since SPIRV code could be executed using Vulkan and OpenCL APIs (*), eventually renaming this to `vulkan.cat` would make sense (not in this PR) .

(*) This is my understanding so far
SPIRV is a language and program written in this language can be executed using Vulkan or OpenCL APIs (similar to how LLVM programs can be executed in x86 or arm hardware). Due to this analogy, I think the following is justified
- having SPRIV events -> justified by the fact we have LLVM events
- having Vulkan events -> justified by the fact we have e.g. x86 events

Our `Compilation` and `Arch` names are misleading. The pass does not really "compile", it actually converts a given type of event into core events. For example, `VisitorTso` does not convert `LlvmXchg` into `TSOXchg` as one would expect from a compilation, but rather into a rmw Load/Store pair (core events).
Following the same principle, we should have only `Vulkan` as a target and a single `VisitorVulkan` which would work both for Vulkan and SPIRV events. AFAIK the current two visitors handle disjoint set of events, so I guess both visitors can be simply unified into one.

Once we add support for OpenCL, we will have OpenCL specific events (or maybe just LLVM events extended with scopes) and a new `VisitorOpenCL` which would convert both SPIRV and OpenCL events into core events.

Basically, there is almost (**) a 1-1 matching between cat/target/visitor.

(**) The exception being vmm, c11 and rc11 models which all share the same visitor + target.

Reply:
Do I understand correctly that you suggest merging VisitorSpirvVulkan and VisitorVulkan into a single class?

Even though it is technically possible, I don't think this would be clean solution. For core events (e.g. plain loads or fences), we need different transformation for vulkan and spirv source language (due to different tags etc). In this case, we either need to know the source language or we will need to introduce new spirv/vulkan events to distinguish between litmus and spirv. To me it looks cleaner to have a different compiler class. Also in general, when you do source code transformations (with whatever other tool) you usually know what is the source language.

The other question is whether we need the middle step in the chain spirv -> vulkan -> core. I don't have a strong opinion here, but for now I would prefer to keep it as it is. This way we have direct mapping from spirv to vulkan events and are more likely to detect potential mistakes. At some point we can drop the vulkan step and transform spirv events directly to core events.

Reply:
I am not really happy that we need this special treatment here and that effectively the spirv compilation works differently that all other compilation visitors (it "really compiled" to vulkan events and then call the vulkan visitor to lower those to core events).

However, I discussed this with Natalia on Friday, and this is closer to a cleaner solution (as proposed in https://github.com/hernanponcedeleon/Dat3M/issues/42). I will take care of this splitting in a different PR, thus I don't see the point in "normalizing" the compilation in this PR with the other ones since that work will eventually be undone.